File: 2019/05/react_day05/src/prob/Computer.js
import React, { useState, useCallback } from "react";
//import React, { useState, useEffect, useContext } from "react";
//import cx from "classnames";
import ProgramGrid from "./ProgramGrid";
import useInterval from "./useInterval";

const MODE_POSITION = 0;
const MODE_IMMEDIATE = 1;

const OP_ADD = 1;
const OP_MULT = 2;
const OP_SAVE = 3;
const OP_WRITE = 4;
const OP_JUMP_IF_TRUE = 5;
const OP_JUMP_IF_FALSE = 6;
const OP_LESS_THAN = 7;
const OP_EQUALS = 8;
const OP_HALT = 99;

// Light on comments.  You'll have to check my python and elixir
// solutions to see the same approach spelled out more clearly!
const digitFromRight = (x, n) => ~~(x / 10 ** n) % 10;
const direct = (memory, pc, n) => memory[pc + n];
const lookup = (memory, pc, n) => {
  const rawInstruction = memory[pc];
  // If instruction is 105, and n=1, mode is the "1", or the 2nd digit
  // from right 0 indexed (3rd when counting naturally)
  const mode = digitFromRight(rawInstruction, n + 1);
  if (mode === MODE_POSITION) {
    const position = direct(memory, pc, n);
    return memory[position];
  } else if (mode === MODE_IMMEDIATE) {
    return direct(memory, pc, n);
  } else {
    throw new Error("Unknown mode");
  }
};

export const Computer = ({ initialMemory, initialInputs }) => {
  const [memory, setMemory] = useState(initialMemory);
  const [inputs, setInputs] = useState(initialInputs);
  const [outputs, setOutputs] = useState([]);
  const [pc, setPC] = useState(0);
  const [halted, setHalted] = useState(false);
  const [autoRun, setAutoRun] = useState(false);

  const step = useCallback(() => {
    if (halted) {
      return;
    }
    const instruction = memory[pc] % 100;
    if (instruction === OP_ADD) {
      const result = lookup(memory, pc, 1) + lookup(memory, pc, 2);
      const outIndex = direct(memory, pc, 3);
      setMemory(oldMemory => {
        let m = [...oldMemory];
        m[outIndex] = result;
        return m;
      });
      setPC(oldPC => oldPC + 4);
    } else if (instruction === OP_MULT) {
      const result = lookup(memory, pc, 1) * lookup(memory, pc, 2);
      const outIndex = direct(memory, pc, 3);
      setMemory(oldMemory => {
        let m = [...oldMemory];
        m[outIndex] = result;
        return m;
      });
      setPC(oldPC => oldPC + 4);
    } else if (instruction === OP_SAVE) {
      // Save: 1 = INPUT
      const thisInput = inputs[0];
      setInputs(oldInputs => {
        let i = [...oldInputs];
        i.shift();
        return i;
      });
      const outIndex = direct(memory, pc, 1);
      setMemory(oldMemory => {
        let m = [...oldMemory];
        m[outIndex] = thisInput;
        return m;
      });
      setPC(oldPC => oldPC + 2);
    } else if (instruction === OP_WRITE) {
      // Write: Output = 1
      const thisOutput = lookup(memory, pc, 1);
      setOutputs(oldOutputs => {
        let m = [...oldOutputs];
        m.push(thisOutput);
        return m;
      });
      setPC(oldPC => oldPC + 2);
    } else if (instruction === OP_JUMP_IF_TRUE) {
      if (lookup(memory, pc, 1) !== 0) {
        setPC(lookup(memory, pc, 2));
      } else {
        setPC(oldPC => oldPC + 3);
      }
    } else if (instruction === OP_JUMP_IF_FALSE) {
      if (lookup(memory, pc, 1) === 0) {
        setPC(lookup(memory, pc, 2));
      } else {
        setPC(oldPC => oldPC + 3);
      }
    } else if (instruction === OP_LESS_THAN) {
      const result = lookup(memory, pc, 1) < lookup(memory, pc, 2) ? 1 : 0;
      const outIndex = direct(memory, pc, 3);
      setMemory(oldMemory => {
        let m = [...oldMemory];
        m[outIndex] = result;
        return m;
      });
      setPC(oldPC => oldPC + 4);
    } else if (instruction === OP_EQUALS) {
      const result = lookup(memory, pc, 1) === lookup(memory, pc, 2) ? 1 : 0;
      const outIndex = direct(memory, pc, 3);
      setMemory(oldMemory => {
        let m = [...oldMemory];
        m[outIndex] = result;
        return m;
      });
      setPC(oldPC => oldPC + 4);
    } else if (instruction === OP_HALT) {
      setHalted(true);
    }
  }, [halted, inputs, memory, pc]);

  const autoStep = useCallback(() => {
    if (autoRun) {
      step();
    }
  }, [autoRun, step]);

  useInterval(autoStep, 100);

  console.log(memory);
  return (
    <div>
      <div>Computer</div>
      <button className="bg-gray-300 rounded p-2" onClick={() => step()}>
        Step
      </button>
      <button
        className="bg-gray-300 rounded p-2 ml-2"
        onClick={() => setAutoRun(true)}
      >
        Auto Run
      </button>
      <ProgramGrid program={memory} current={pc} />
      <div className="bg-green-200 max-w-md mx-auto rounded p-2">
        Outputs:
        {outputs.map(o => (
          <span className="mx-2">{o}</span>
        ))}
      </div>
    </div>
  );
};
export default Computer;

File: 2019/07/elixir_day07/lib/computer.ex
defmodule ElixirDay07.Computer do
  defstruct [
    :memory,
    :pc,
    :inputs,
    :outputs,
    :halted
  ]

  @op_add 1
  @op_mult 2
  @op_save 3
  @op_write 4
  @op_jump_if_true 5
  @op_jump_if_false 6
  @op_less_than 7
  @op_equals 8
  @op_halt 99

  @mode_position 0
  @mode_immediate 1

  alias ElixirDay07.Computer

  def solve(memory, inputs) when is_list(memory) and is_list(inputs) do
    Computer.new(memory, inputs)
    |> Computer.execute()
    |> Computer.outputs()
  end

  def new(memory, inputs) when is_list(memory) and is_list(inputs) do
    %Computer{
      memory: Array.from_list(memory),
      inputs: inputs,
      outputs: [],
      pc: 0,
      halted: false
    }
  end

  # Get the direct value of the memory address of the Nth arg, or PC + N
  def direct(%Computer{pc: pc, memory: memory}, n) do
    Array.get(memory, pc + n)
  end

  # Get the dereferenced value of the Nth arg, after checking the Nth mode of the current instruction.
  def lookup(%Computer{pc: pc, memory: memory} = computer, n) do
    raw_instruction = Array.get(memory, pc)
    # If instruction is 105, and n=1, mode is the "1", or the 2nd digit
    # from right 0 indexed (3rd when counting naturally)
    mode = digit_from_right(raw_instruction, n + 1)

    case mode do
      @mode_position ->
        position = direct(computer, n)
        Array.get(memory, position)

      @mode_immediate ->
        direct(computer, n)

      _ ->
        raise "Unknown mode"
    end
  end

  def execute(computer) do
    computer = execute_step(computer)

    if computer.halted do
      computer
    else
      execute(computer)
    end
  end

  def outputs(%Computer{outputs: outputs}), do: outputs

  def execute_step(%Computer{pc: pc, memory: memory} = c) do
    instruction = Array.get(memory, pc) |> rem(100)
    do_execute_step(c, instruction)
  end

  # ADD: 3 = 1 + 2
  def do_execute_step(%Computer{pc: pc, memory: memory} = c, @op_add) do
    result = lookup(c, 1) + lookup(c, 2)
    new_memory = Array.set(memory, direct(c, 3), result)
    new_pc = pc + 4
    %Computer{c | memory: new_memory, pc: new_pc}
  end

  # MULT: 3 = 1 * 2
  def do_execute_step(%Computer{pc: pc, memory: memory} = c, @op_mult) do
    result = lookup(c, 1) * lookup(c, 2)
    new_memory = Array.set(memory, direct(c, 3), result)
    new_pc = pc + 4
    %Computer{c | memory: new_memory, pc: new_pc}
  end

  # SAVE: 1 = Input
  def do_execute_step(%Computer{pc: pc, memory: memory, inputs: inputs} = c, @op_save) do
    [this_input | new_inputs] = inputs
    new_memory = Array.set(memory, direct(c, 1), this_input)
    new_pc = pc + 2
    %Computer{c | memory: new_memory, pc: new_pc, inputs: new_inputs}
  end

  # WRITE: Output = 1
  def do_execute_step(%Computer{pc: pc, outputs: outputs} = c, @op_write) do
    this_output = lookup(c, 1)
    new_outputs = outputs ++ [this_output]
    new_pc = pc + 2
    %Computer{c | pc: new_pc, outputs: new_outputs}
  end

  # HALT: Stop
  def do_execute_step(c, @op_halt) do
    %Computer{c | halted: true}
  end

  def do_execute_step(%Computer{pc: pc} = c, @op_jump_if_true) do
    new_pc =
      if lookup(c, 1) != 0 do
        lookup(c, 2)
      else
        pc + 3
      end

    %Computer{c | pc: new_pc}
  end

  def do_execute_step(%Computer{pc: pc} = c, @op_jump_if_false) do
    new_pc =
      if lookup(c, 1) == 0 do
        lookup(c, 2)
      else
        pc + 3
      end

    %Computer{c | pc: new_pc}
  end

  def do_execute_step(%Computer{pc: pc, memory: memory} = c, @op_less_than) do
    result = if lookup(c, 1) < lookup(c, 2), do: 1, else: 0
    new_memory = Array.set(memory, direct(c, 3), result)
    new_pc = pc + 4
    %Computer{c | memory: new_memory, pc: new_pc}
  end

  def do_execute_step(%Computer{pc: pc, memory: memory} = c, @op_equals) do
    result = if lookup(c, 1) == lookup(c, 2), do: 1, else: 0
    new_memory = Array.set(memory, direct(c, 3), result)
    new_pc = pc + 4
    %Computer{c | memory: new_memory, pc: new_pc}
  end

  def do_execute_step(_, opcode) do
    raise "Unknown opcode " <> Integer.to_string(opcode)
  end

  def digit_from_right(x, n) do
    x
    |> Integer.floor_div(round(:math.pow(10, n)))
    |> rem(10)
  end
end
